File: blink-windows/src/Blink.Core.Tests/AdaptiveTimingEngineTests.cs
using Blink.Core.Timer;

namespace Blink.Core.Tests;

public class AdaptiveTimingEngineTests
{
    [Fact]
    public void NotEnoughData_NoSuggestion()
    {
        var engine = new AdaptiveTimingEngine();
        for (int i = 0; i < 5; i++)
            engine.RecordAcceptedBreak(1200);
        Assert.Null(engine.SuggestedDuration());
    }

    [Fact]
    public void EnoughData_MedianSuggestion()
    {
        var engine = new AdaptiveTimingEngine();
        foreach (var i in new[] { 1200, 1300, 1100, 1400, 1250, 1350, 1150, 1500, 1200, 1300 })
            engine.RecordAcceptedBreak(i);
        var suggested = engine.SuggestedDuration();
        Assert.NotNull(suggested);
        Assert.InRange(suggested.Value, 900, 2700);
    }

    [Fact]
    public void ClampsToMin()
    {
        var engine = new AdaptiveTimingEngine();
        for (int i = 0; i < 15; i++)
            engine.RecordAcceptedBreak(300); // 5 min
        Assert.Equal(900, engine.SuggestedDuration());
    }

    [Fact]
    public void ClampsToMax()
    {
        var engine = new AdaptiveTimingEngine();
        for (int i = 0; i < 15; i++)
            engine.RecordAcceptedBreak(5000); // 83 min
        Assert.Equal(2700, engine.SuggestedDuration());
    }

    [Fact]
    public void LoadSave_Roundtrips()
    {
        var engine = new AdaptiveTimingEngine();
        for (int i = 0; i < 12; i++)
            engine.RecordAcceptedBreak(1200 + i * 10);
        var saved = engine.SavedIntervals();

        var engine2 = new AdaptiveTimingEngine();
        engine2.Load(saved);
        Assert.Equal(engine.SuggestedDuration(), engine2.SuggestedDuration());
    }

    [Fact]
    public void BufferCapsAt50()
    {
        var engine = new AdaptiveTimingEngine();
        for (int i = 0; i < 60; i++)
            engine.RecordAcceptedBreak(1200 + i);
        Assert.Equal(50, engine.SavedIntervals().Length);
    }
}

File: blink-macos/BlinkCore/Tests/BlinkCoreTests/AdaptiveTimingEngineTests.swift
import Foundation
import Testing
@testable import BlinkCore

@Suite("AdaptiveTimingEngine")
struct AdaptiveTimingEngineTests {
    @Test("Not enough data = no suggestion")
    func notEnoughData() {
        let engine = AdaptiveTimingEngine()
        for _ in 0..<5 {
            engine.recordAcceptedBreak(intervalSinceLastBreak: 1200)
        }
        #expect(engine.suggestedDuration() == nil, "Need 10+ data points")
    }

    @Test("10+ data points = median suggestion")
    func enoughData() {
        let engine = AdaptiveTimingEngine()
        let intervals: [TimeInterval] = [1200, 1300, 1100, 1400, 1250, 1350, 1150, 1500, 1200, 1300]
        for i in intervals {
            engine.recordAcceptedBreak(intervalSinceLastBreak: i)
        }
        let suggested = engine.suggestedDuration()
        #expect(suggested != nil)
        #expect(suggested! >= 900, "Should be at least 15 min")
        #expect(suggested! <= 2700, "Should be at most 45 min")
    }

    @Test("Clamps to min/max")
    func clamping() {
        let engine = AdaptiveTimingEngine()
        // All very short intervals
        for _ in 0..<15 {
            engine.recordAcceptedBreak(intervalSinceLastBreak: 300) // 5 min
        }
        #expect(engine.suggestedDuration() == 900, "Should clamp to 15 min minimum")

        let engine2 = AdaptiveTimingEngine()
        for _ in 0..<15 {
            engine2.recordAcceptedBreak(intervalSinceLastBreak: 5000) // 83 min
        }
        #expect(engine2.suggestedDuration() == 2700, "Should clamp to 45 min maximum")
    }

    @Test("Load and save round-trips")
    func loadSave() {
        let engine = AdaptiveTimingEngine()
        for i in 0..<12 {
            engine.recordAcceptedBreak(intervalSinceLastBreak: 1200 + Double(i) * 10)
        }
        let saved = engine.savedIntervals()

        let engine2 = AdaptiveTimingEngine()
        engine2.load(intervals: saved)
        #expect(engine2.suggestedDuration() == engine.suggestedDuration())
    }

    @Test("Buffer caps at 50 entries")
    func bufferCap() {
        let engine = AdaptiveTimingEngine()
        for i in 0..<60 {
            engine.recordAcceptedBreak(intervalSinceLastBreak: 1200 + Double(i))
        }
        #expect(engine.savedIntervals().count == 50)
    }
}
